src/operators/products.c

```diff
@@ -0,0 +1,217 @@
+/*
+   IGraph library.
+   Copyright (C) 2025  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_operators.h"
+
+#include "igraph_conversion.h"
+#include "igraph_interface.h"
+
+#include "core/interruption.h"
+#include "math/safe_intop.h"
+
+
+static igraph_error_t cartesian_product(igraph_t *res,
+   const igraph_t *g1, const igraph_t *g2) {
+      
+   igraph_bool_t directed1 = igraph_is_directed(g1);
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Cartesian product of directed and undirected graphs is not supported. "
+                   "You can convert an undirected graph to a directed one using igraph_to_directed, "
+                   "or a directed graph to an undirected one using igraph_to_undirected.", IGRAPH_EINVAL);
+   }
+  
+   igraph_bool_t directed = directed1 && directed2;
+
+   igraph_integer_t vg1 = igraph_vcount(g1);
+   igraph_integer_t vg2 = igraph_vcount(g2);
+   igraph_integer_t eg1 = igraph_ecount(g1);
+   igraph_integer_t eg2 = igraph_ecount(g2);
+   igraph_integer_t vres;
+   igraph_integer_t eres;
+   igraph_integer_t temp;
+
+   IGRAPH_SAFE_MULT(vg1, vg2, &vres);
+   igraph_vector_int_t edges;
+
+   // new edge count = vg1*e2 + vg2*e1
+   IGRAPH_SAFE_MULT(vg1, eg2, &eres);
+   IGRAPH_SAFE_MULT(vg2, eg1, &temp);
+   IGRAPH_SAFE_ADD(eres, temp, &eres);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 2*eres);
+
+   igraph_integer_t from, to;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vg2 + j
+
+   igraph_integer_t edge_index = 0;
+   // Edges from g1
+   for (i = 0; i < eg1; ++i) {
+      igraph_edge(g1, i, &from, &to);
```

Comment:
Avoid this function and use the `IGRAPH_FROM` and `IGRAPH_TO` macros instead. They have better performance as they avoid function calls.